lib/ecto/adapters/sql.ex

```diff
@@ -38,19 +38,15 @@ defmodule Ecto.Adapters.SQL do
       ## Types
 
       @doc false
-      def autogenerate(:id), do: nil
-      def autogenerate(:embed_id), do: Ecto.UUID.autogenerate()
-      def autogenerate(:binary_id), do: Ecto.UUID.autogenerate()
+      def autogenerate(:id),        do: nil
+      def autogenerate(:embed_id),  do: Ecto.UUID.generate()
```

Comment:
I see. The autogenerate API is still a bit everywhere unfortunately. autogenerate for embeds generate an Ecto value, for binary ids the DB value and the autogenerate API Ecto values.

I don't mind much about the autogenerate in adapters bit but the one in custom types is bothering me a lot because of the amount of unecessary paths. Here is what is happening for UUID:

  1. autogenerate calls "bingenerate" which calls "generate"
  2. ecto then dumps the type converting "generate" back into "bingenerate"
  3. the type is loaded after the record is modified, going from "bingenerate" to "generate"

So we have:

    bingenerate -> generate -> bingenerate -> generate

Previously we had:

  1. autogenerate calls "bingenerate"
  3. the type is loaded after the record is modified, going from "bingenerate" to "generate"

I think we can at least skip step 3 though by storing those values. I will give it a try. It is still one extra pass compared to the ideal case but I don't think we can avoid it if we want to keep the adapters responsible for what they are doing.